crates/stateless/src/validation.rs

```diff
@@ -171,16 +212,241 @@ pub fn stateless_validation_with_trie<T, ChainSpec, E>(
+    public_keys: Vec<UncompressedPublicKey>,
+    witness: FlatExecutionWitness,
+    chain_spec: Arc<ChainSpec>,
+    evm_config: E,
+) -> Result<(B256, HashedPostState), StatelessValidationError>
+where
+    ChainSpec: Send + Sync + EthChainSpec<Header = Header> + EthereumHardforks + Debug,
+    E: ConfigureEvm<Primitives = EthPrimitives> + Clone + 'static,
+{
+    let recovered_block = recover_block_with_public_keys(current_block, public_keys, &*chain_spec)?;
+
+    let hash = keccak256(&witness.parent_header);
+    let parent_header = alloy_rlp::decode_exact::<Header>(&witness.parent_header)
+        .map(|h| SealedHeader::new(h, hash))
+        .map_err(|_| StatelessValidationError::HeaderDeserializationFailed)?;
+    let db = witness.create_db();
+
+    let hashed_post_state = stateless_validation_execution(
+        &recovered_block,
+        &parent_header,
+        chain_spec,
+        evm_config,
+        db,
+    )?;
+
+    Ok((recovered_block.hash_slow(), hashed_post_state))
+}
+
+/// Executes the block statelessly with a provided `Database` and performs post-execution
+/// validation.
+fn stateless_validation_execution<DB: reth_evm::Database, ChainSpec, E>(
+    current_block: &RecoveredBlock<Block>,
+    parent: &SealedHeader,
+    chain_spec: Arc<ChainSpec>,
+    evm_config: E,
+    db: DB,
+) -> Result<HashedPostState, StatelessValidationError>
+where
+    ChainSpec: Send + Sync + EthChainSpec<Header = Header> + EthereumHardforks + Debug,
+    E: ConfigureEvm<Primitives = EthPrimitives> + Clone + 'static,
+{
+    // Validate block against pre-execution consensus rules
+    validate_block_consensus(chain_spec.clone(), current_block, parent)?;
+
+    // Execute the block
+    let executor = evm_config.executor(db);
+    let output = track_cycles!(
+        "block_execution",
+        executor
+            .execute(current_block)
+            .map_err(|e| StatelessValidationError::StatelessExecutionFailed(e.to_string()))?
+    );
+
+    // Post validation checks
+    validate_block_post_execution(current_block, &chain_spec, &output.receipts, &output.requests)
+        .map_err(StatelessValidationError::ConsensusValidationFailed)?;
+
+    Ok(HashedPostState::from_bundle_state::<KeccakKeyHasher>(&output.state.state))
+}
+
+/// Cross-validates flatdb state against a cryptographically verified sparse trie witness.
+///
+/// Flatdb provides optimized flat storage for EVM execution, but lacks cryptographic
+/// verification. This function ensures flatdb was correctly constructed by checking that:
+/// - Block hashes match between flatdb and the verified trie witness
+/// - Contract bytecode hashes are correct
+/// - Account state matches the sparse trie (witness may contain extra accounts; that's ok)
+/// - Post-state root computed from flatdb matches the block header
+///
+/// Returns block hash if all validations pass.
+pub fn stateless_validation_flatdb_storage_check<T>(
+    current_block: Block,
+    trie_witness: ExecutionWitness,
```

Comment:
We receive the usual `ExecutionWitness`, since we need to cryptographically validate pre-state. This will be our source of truth for state.